### DISCUSSION TITLE
Upstash adapter not working for middleware

### DISCUSSION THREAD

**Discussion by maitrungduc1410** [nextauthjs/next-auth#7034] [Help] [answered]
Hi all,

I'm using Upstash Redis Adapter.

My app has 2 routes: `/` and `/classes`

And I have a middleware to protect `/classes`:
```ts
export { default } from "next-auth/middleware"

export const config = {
  matcher: ["/classes"],
}
```

When an authenticated user access `/` -> redirect to my own provider to perform SSO -> redirect back to my NextJS app, session is stored on Upstash successfully

With the same authenticated user (already login), If I try to access the app at `/classes` , it just shows the login page (click the login button just refresh the login page, not working):

<img width="558" alt="Screenshot 2023-03-22 at 10 34 44 PM" src="https://user-images.githubusercontent.com/16630193/226964156-da10b371-ef88-401d-a313-f33854125a7a.png">

I tried using Incognito tab, and directly access `/classes`, the SSO flow is successful, session stored on Upstash successful, but the Login page is still showing.

And this does not happen if I remove `adapter` in my `authOptions`.

Please help, Thanks

My environment:

- node: v16.18.0
- next: 13.2.4
- next-auth: 4.20.1
- @next-auth/upstash-redis-adapter: "^3.0.4",
- @upstash/redis: "^1.20.1",

**Comment by maitrungduc1410** [comment]
I tried to add custom logic to my middleware:
```ts
import { withAuth } from "next-auth/middleware"

export default withAuth(
  // `withAuth` augments your `Request` with the user's token.
  function middleware(req) {
    console.log(req.nextauth.token)
  },
  {
    callbacks: {
      authorized: ({ token, req }) => {
        console.log(token)
        console.log(req.headers)
        return true
      },
    },
  }
)

export const config = { matcher: ["/classes"] }
```
And figured out that `token` is null for the 2 console log, even there's `next-auth.session-token` in `req.headers`

**Reply by maitrungduc1410** [reply]
I tried to get `next-auth.session-token` then query to Redis to get session, but wasn't successful because its Edge Runtime in Middleware (https://nextjs.org/docs/api-reference/edge-runtime), and it's very limited, we can't access backend from here

**Reply by chronark** [reply]
upstash redis is available over http and works in edge functions. you can use `@upstash/redis` for example

**Reply by maitrungduc1410** [reply]
Sorry, my bad, yes Upstash works fine because it uses http (from my local to Upstash server)

I was trying to switch to my local Redis instance instead if Upstash

And because in Edge Runtime ,we have limited access so it doesn't work.

Sorry for the confusion

**Comment by chronark** [comment]
Hey, is your code open source?

How can I reproduce it?

**Reply by maitrungduc1410** [reply]
my code is private, 

as I mentioned, my code is very basic, just follow docs of Next-auth to setup Upstash: https://authjs.dev/reference/adapter/upstash-redis

then create middleware to protect a page with Next-Auth middleware: https://next-auth.js.org/tutorials/securing-pages-and-api-routes#nextjs-middleware

**Comment by chronark** [comment]
> And figured out that token is null for the 2 console log, even there's next-auth.session-token in req.headers

`token` is null, because you're using sessions, instead of `jwts` I think. You can only do either one

**Comment by chronark** [comment, marked as answer]
```ts
import { withAuth } from "next-auth/middleware"

export default withAuth(
  // `withAuth` augments your `Request` with the user's token.
  function middleware(req) {
    console.log(req.nextauth.token)
  },
  {
    callbacks: {
      authorized: ({req}) => {
        const parsed = new RegExp(/next-auth.session-token=([0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12})/gm).exec(req.headers.get("cookie"))
        const sessionToken = parsed.at(1)
        console.log({sessionToken})
        return true
      },
    },
  }
)

export const config = { matcher: ["/classes"] }

```

This worked and I could look up the session in redis
```ts
type Session = {
 userId: string,
 // ...
}
const session = await redis.get<Session>(`user:session:${sessionToken}`)
```
